### DISCUSSION TITLE
Debugging on nucleo_u575zi_q

### DISCUSSION THREAD

**Discussion by VineetaNarkhede-eaton** [zephyrproject-rtos/zephyr#45167] [Q&A]
Hi,

I am working on flashing and debugging "Hello world" example code on nucleo_u575zi_q, following this document, https://docs.zephyrproject.org/3.0.0/boards/arm/nucleo_u575zi_q/doc/nucleou575zi_q.html#programming
As per document STM32U5 is not currently supported in openocd and temporary workaround given. However, I downloaded OpenOCD-20211118-0.11.0, and ran west debug command which looks to be working fine. openocd configuration file is also available for STM32U5 at https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/arm/nucleo_u575zi_q/support/openocd.cfg

Question: is the debugging section in this document updated https://docs.zephyrproject.org/3.0.0/boards/arm/nucleo_u575zi_q/doc/nucleou575zi_q.html, it looks like openocd support is available for STM32U5, please confirm.

```
west flash -r openocd
west debug -r openocd
```
**Logs:**
```
C:\Users\User1\zephyrsetup\zephyrproject_3_0_0\zephyr>west debug -r openocd
-- west debug: rebuilding
ninja: no work to do.
-- west debug: using runner openocd
-- runners.openocd: OpenOCD GDB server running on port 3333; no thread info available
Open On-Chip Debugger 0.11.0 (2021-11-18) [https://github.com/sysprogs/openocd]
Licensed under GNU GPL v2
libusb1 09e75e98b4d9ea7909e8837b7a3f00dda4589dc3
For bug reports, read
        http://openocd.org/doc/doxygen/bugs.html
Info : DEPRECATED target event trace-config; use TPIU events {pre,post}-{enable,disable}
Info : STLINK V3J7M3 (API v3) VID:PID 0483:374E
Info : Target voltage: 3.302796
Info : Unable to match requested speed 500 kHz, using 200 kHz
Info : Unable to match requested speed 500 kHz, using 200 kHz
Info : clock speed 200 kHz
Info : stlink_dap_op_connect(connect)
Info : SWD DPIDR 0x0be12477
Info : STM32U575ZITxQ.cpu: Cortex-M33 r0p4 processor detected
Info : STM32U575ZITxQ.cpu: target has 8 breakpoints, 4 watchpoints
Info : STM32U575ZITxQ.cpu: external reset detected
Info : starting gdb server for STM32U575ZITxQ.cpu on 3333
Info : Listening on port 3333 for gdb connections
    TargetName         Type       Endian TapName            State
--  ------------------ ---------- ------ ------------------ ------------
 0* STM32U575ZITxQ.cpu cortex_m   little STM32U575ZITxQ.cpu reset

Info : Listening on port 6666 for tcl connections
Info : Listening on port 4444 for telnet connections
C:\softwares\gnu_arm_embedded\bin\arm-none-eabi-gdb.exe: warning: Couldn't determine a path for the index cache directory.
GNU gdb (GNU Arm Embedded Toolchain 10.3-2021.07) 10.2.90.20210621-git
Copyright (C) 2021 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "--host=i686-w64-mingw32 --target=arm-none-eabi".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from C:/Users/User1/zephyrsetup/zephyrproject_3_0_0/zephyr/build/zephyr/zephyr.elf...
RInefom :o atcece pdteibngu g'ggdib'n cgo nnuesctiionng o n :t3cp3/333333

Debugger attaching: halting execution
Info : Unable to match requested speed 480 kHz, using 200 kHz
Info : Unable to match requested speed 480 kHz, using 200 kHz
CPU in Non-Secure state
target halted due to debug-request, current mode: Thread
xPSR: 0xf9000000 pc: 0xfffffffe msp: 0xfffffffc
force hard breakpoints
Info : device idcode = 0x20016482 (STM32U57/U58xx - Rev 'unknown' : 0x2001)
Info : TZEN = 0 : TrustZone disabled by option bytes
Info : RDP level 0 (0xAA)
Info : flash size = 2048kbytes
Info : flash mode : dual-bank
Info : device idcode = 0x20016482 (STM32U57/U58xx - Rev 'unknown' : 0x2001)
Info : TZEN = 0 : TrustZone disabled by option bytes
Info : RDP level 0 (0xAA)
Info : flash size = 2048kbytes
Info : flash mode : dual-bank
Info : device idcode = 0x20016482 (STM32U57/U58xx - Rev 'unknown' : 0x2001)
Info : TZEN = 0 : TrustZone disabled by option bytes
Info : RDP level 0 (0xAA)
Info : OTP size is 512 bytes, base address is 0x0bfa0000
Error: Failed to read memory at 0xfffffffe
0xfffffffe in ?? ()
TZEN option bit disabled
workarea_addr 536870912
Loading section rom_start, size 0x234 lma 0x8000000
Loading section text, size 0x3478 lma 0x8000234
Loading section .ARM.exidx, size 0x8 lma 0x80036ac
Loading section initlevel, size 0x80 lma 0x80036b4
Loading section devices, size 0x120 lma 0x8003734
Loading section sw_isr_table, size 0x3e8 lma 0x8003854
Loading section device_handles, size 0x70 lma 0x8003c3c
Loading section rodata, size 0x2e4 lma 0x8003cac
Loading section datas, size 0x34 lma 0x8003f90
Loading section device_states, size 0x30 lma 0x8003fc4
Info : Padding image section 0 at 0x08003ff4 with 12 bytes (bank write end alignment)
Warn : keep_alive() was not invoked in the 1000 ms timelimit. GDB alive packet not sent! (1065 ms). Workaround: increase "set remotetimeout" in GDB
Info : Unable to match requested speed 480 kHz, using 200 kHz
Info : Unable to match requested speed 480 kHz, using 200 kHz
target halted due to debug-request, current mode: Thread
xPSR: 0xf9000000 pc: 0x0800148c msp: 0x200006e0
Start address 0x0800148c, load size 16372
--Type <RET> for more, q to quit, c to continue without paging--c
Transfer rate: 6 KB/sec, 1637 bytes/write.
(gdb) Info : STM32U575ZITxQ.cpu: external reset detected

(gdb)
```

**Comment by erwango** [comment]
Yes, this section is up to date.
Please note the lines before these instructiosn, that mention that you should use ST openocd fork and build your sample by pointing to this openocd version:
`west build -b b_u585i_iot02a samples/basic/blinky -- -DOPENOCD="<path_to_openocd>/openocd/src/openocd" -DOPENOCD_DEFAULT_PATH="<path_to_openocd>/openocd/tcl/"`

**Reply by VineetaNarkhede-eaton** [reply]
Thanks for your response @erwango 
You have mentioned **b_u585i_iot02a**, my question is for **nucleo_u575zi_q** target https://docs.zephyrproject.org/3.0.0/boards/arm/nucleo_u575zi_q/doc/nucleou575zi_q.html
Please confirm if same applies to nucleo_u575zi_q target also. 

Few questions:
1. west debug is working with OpenOCD-20211118-0.11.0 then what is the reason for compiling and using "openocd-cubeide-r3" as given in document?
2. Why this debugging solution is mentioned as a temporary workaround,
![image](https://user-images.githubusercontent.com/88422239/165703300-4b4e1dbd-9e1c-447c-8d88-2f49b2d4c9c3.png)
3. I am following instructions given in readme file of "openocd-cubeide-r3" and stuck at configure step. c compiler is unable to create executable. Any suggestion to solve this problem?
![image](https://user-images.githubusercontent.com/88422239/165704009-e0470ab0-8b31-464c-a036-695ed57a2b9e.png)

**Reply by erwango** [reply]
This temporary workaround is provided as default openocd version provided with Zephyr SDK does not support U5 (which explains the issue you're seeing when trying to use it). Once U5 support will be upstreamed in an openocd version available within Zephyr SDK, it won't be required anymore.
Regarding the build instructions, I won't be able to help as I'm not a windows user and have not faced this issue. I'd suggest to look on the web how to build openocd on windows. Alternative it to get openocd executable from a recent CubeIDE package (Though, I can't provide instructions on how to do that)

**Reply by VineetaNarkhede-eaton** [reply]
@erwango I am using zephyr version 3.0.0 on windows, I have installed GNU Arm Embedded toolchain as per https://docs.zephyrproject.org/3.0.0/getting_started/index.html#install-a-toolchain
**not using zephyr SDK**
![image](https://user-images.githubusercontent.com/88422239/166222419-79eb5f1c-82e9-4033-86b7-c9b9b4adbf74.png)

and using latest executable of openocd https://gnutoolchains.com/arm-eabi/openocd/
OpenOCD-20211118-0.11.0\bin\openocd.exe

**Reply by erwango** [reply]
As already mentioned:

> I won't be able to help as I'm not a windows user and have not faced this issue. I'd suggest to look on the web how to build openocd on windows. Alternative it to get openocd executable from a recent CubeIDE package (Though, I can't provide instructions on how to do that)

One alternative idea is to use CubeIDE to debug you zephyr.bin:
To do this you need to set up a CubeMx project using the requested board, build and debug to see that this configuration is right and works with your board. Once this is ok, get CubeIDE to flash and debug zephyr binary (this could be done by updating cubeide path to the binary to use).
